src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,18 +358,59 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
-runHook = (hookName, args) ->
+class Hook
+	constructor: (hookFn) ->
+		@hookFn = hookFn
+		@executed = false
+
+	run: (args...) ->
+		Promise.try =>
+			@hookFn(args...)
+		.finally =>
+			@executed = true
+
+class SideEffectHook extends Hook
+	constructor: (hookFn) ->
+		@rollbackFns = []
+		@rolledBack = false
+		super(hookFn)
+
+	registerRollback: (fn) ->
```

Comment:
You should add a `return` at the end of this function, right now it can return either a promise or a number and it's possible that will cause the promise to be waited upon by the parent in some cases, eg
```coffee
Promise.try =>
	@registerRollback(fn)
```
will sometimes wait on the fn and sometimes not